File: apple/AestheticAlgorithm/Algorithms/Sorting/MergeSort.swift
import Foundation

struct MergeSort: AlgorithmExecutable {
    let id = "merge-sort"
    let name = "Merge Sort"
    let category = "sorting"
    let timeComplexity = (best: "O(n log n)", average: "O(n log n)", worst: "O(n log n)")
    let spaceComplexity = "O(n)"

    let pseudocode: [(line: Int, text: String)] = [
        (0, "procedure mergeSort(A, left, right)"),
        (1, "  if left < right then"),
        (2, "    mid ← (left + right) / 2"),
        (3, "    mergeSort(A, left, mid)"),
        (4, "    mergeSort(A, mid+1, right)"),
        (5, "    merge(A, left, mid, right)"),
        (6, ""),
        (7, "procedure merge(A, left, mid, right)"),
        (8, "  create temp arrays L, R"),
        (9, "  while i < len(L) and j < len(R) do"),
        (10, "    if L[i] <= R[j] then"),
        (11, "      A[k] ← L[i]; i ← i + 1"),
        (12, "    else"),
        (13, "      A[k] ← R[j]; j ← j + 1"),
        (14, "    k ← k + 1"),
        (15, "  copy remaining elements"),
        (16, "  return A")
    ]

    func generateSteps(from input: [Int]) -> [AlgorithmStep] {
        var array = input
        var steps: [AlgorithmStep] = []
        var sortedIndices: [Int] = []

        if array.isEmpty { return steps }

        // Initial state
        steps.append(AlgorithmStep(
            type: .highlight,
            array: array,
            highlightedIndices: [],
            secondaryIndices: [],
            sortedIndices: [],
            pseudocodeLine: 0,
            description: "Starting merge sort on array of \(array.count) elements"
        ))

        mergeSort(&array, left: 0, right: array.count - 1, steps: &steps, sortedIndices: &sortedIndices)

        // Final sorted state
        steps.append(AlgorithmStep(
            type: .sorted,
            array: array,
            highlightedIndices: Array(0..<array.count),
            secondaryIndices: [],
            sortedIndices: Array(0..<array.count),
            pseudocodeLine: 16,
            description: "Merge sort complete"
        ))

        return steps
    }

    private func mergeSort(_ array: inout [Int], left: Int, right: Int, steps: inout [AlgorithmStep], sortedIndices: inout [Int]) {
        guard left < right else { return }

        let mid = (left + right) / 2

        // Split step
        steps.append(AlgorithmStep(
            type: .highlight,
            array: array,
            highlightedIndices: Array(left...mid),
            secondaryIndices: Array((mid + 1)...right),
            sortedIndices: sortedIndices,
            pseudocodeLine: 2,
            description: "Splitting [\(left)...\(right)] at midpoint \(mid)"
        ))

        // Recurse left half
        mergeSort(&array, left: left, right: mid, steps: &steps, sortedIndices: &sortedIndices)

        // Recurse right half
        mergeSort(&array, left: mid + 1, right: right, steps: &steps, sortedIndices: &sortedIndices)

        // Merge the two halves
        merge(&array, left: left, mid: mid, right: right, steps: &steps, sortedIndices: &sortedIndices)
    }

    private func merge(_ array: inout [Int], left: Int, mid: Int, right: Int, steps: inout [AlgorithmStep], sortedIndices: inout [Int]) {
        let leftArray = Array(array[left...mid])
        let rightArray = Array(array[(mid + 1)...right])

        // Show merge start
        steps.append(AlgorithmStep(
            type: .highlight,
            array: array,
            highlightedIndices: Array(left...mid),
            secondaryIndices: Array((mid + 1)...right),
            sortedIndices: sortedIndices,
            pseudocodeLine: 8,
            description: "Merging [\(left)...\(mid)] and [\(mid + 1)...\(right)]"
        ))

        var i = 0
        var j = 0
        var k = left

        while i < leftArray.count && j < rightArray.count {
            // Compare step
            steps.append(AlgorithmStep(
                type: .compare,
                array: array,
                highlightedIndices: [k],
                secondaryIndices: [left + i, mid + 1 + j].filter { $0 <= right },
                sortedIndices: sortedIndices,
                pseudocodeLine: 10,
                description: "Comparing \(leftArray[i]) and \(rightArray[j])"
            ))

            if leftArray[i] <= rightArray[j] {
                array[k] = leftArray[i]
                steps.append(AlgorithmStep(
                    type: .insert,
                    array: array,
                    highlightedIndices: [k],
                    secondaryIndices: [],
                    sortedIndices: sortedIndices,
                    pseudocodeLine: 11,
                    description: "Placing \(leftArray[i]) from left subarray at position \(k)"
                ))
                i += 1
            } else {
                array[k] = rightArray[j]
                steps.append(AlgorithmStep(
                    type: .insert,
                    array: array,
                    highlightedIndices: [k],
                    secondaryIndices: [],
                    sortedIndices: sortedIndices,
                    pseudocodeLine: 13,
                    description: "Placing \(rightArray[j]) from right subarray at position \(k)"
                ))
                j += 1
            }
            k += 1
        }

        // Copy remaining left elements
        while i < leftArray.count {
            array[k] = leftArray[i]
            steps.append(AlgorithmStep(
                type: .insert,
                array: array,
                highlightedIndices: [k],
                secondaryIndices: [],
                sortedIndices: sortedIndices,
                pseudocodeLine: 15,
                description: "Copying remaining element \(leftArray[i]) from left subarray"
            ))
            i += 1
            k += 1
        }

        // Copy remaining right elements
        while j < rightArray.count {
            array[k] = rightArray[j]
            steps.append(AlgorithmStep(
                type: .insert,
                array: array,
                highlightedIndices: [k],
                secondaryIndices: [],
                sortedIndices: sortedIndices,
                pseudocodeLine: 15,
                description: "Copying remaining element \(rightArray[j]) from right subarray"
            ))
            j += 1
            k += 1
        }

        // Mark merged region as sorted if this is the final merge
        if left == 0 && right == array.count - 1 {
            sortedIndices = Array(0..<array.count)
        }

        // Show merged result
        steps.append(AlgorithmStep(
            type: .sorted,
            array: array,
            highlightedIndices: Array(left...right),
            secondaryIndices: [],
            sortedIndices: sortedIndices,
            pseudocodeLine: 5,
            description: "Merged region [\(left)...\(right)]: \(Array(array[left...right]))"
        ))
    }
}

File: web/src/algorithms/sorting/merge-sort.ts
import type { AlgorithmImplementation, AlgorithmStep } from '@/lib/types/algorithm';

export const mergeSort: AlgorithmImplementation = {
  id: 'merge-sort',
  name: 'Merge Sort',
  category: 'sorting',
  timeComplexity: { best: 'O(n log n)', average: 'O(n log n)', worst: 'O(n log n)' },
  spaceComplexity: 'O(n)',
  pseudocode: [
    { line: 0, text: 'procedure mergeSort(A, left, right)' },
    { line: 1, text: '  if left < right then' },
    { line: 2, text: '    mid \u2190 (left + right) / 2' },
    { line: 3, text: '    mergeSort(A, left, mid)' },
    { line: 4, text: '    mergeSort(A, mid+1, right)' },
    { line: 5, text: '    merge(A, left, mid, right)' },
    { line: 6, text: 'procedure merge(A, left, mid, right)' },
    { line: 7, text: '  while i \u2264 mid and j \u2264 right do' },
    { line: 8, text: '    if A[i] \u2264 A[j] then copy A[i]' },
    { line: 9, text: '    else copy A[j]' },
    { line: 10, text: '  copy remaining elements' },
  ],

  generateSteps(input: number[]): AlgorithmStep[] {
    const array = [...input];
    const steps: AlgorithmStep[] = [];
    const sortedIndices: number[] = [];

    function mergeSortRecursive(left: number, right: number): void {
      if (left >= right) return;

      const mid = Math.floor((left + right) / 2);

      // Show the split
      steps.push({
        type: 'select',
        array: [...array],
        highlightedIndices: [mid],
        secondaryIndices: Array.from({ length: right - left + 1 }, (_, i) => left + i),
        sortedIndices: [...sortedIndices],
        pseudocodeLine: 2,
        description: `Split array at index ${mid}: [${left}..${mid}] and [${mid + 1}..${right}]`,
      });

      mergeSortRecursive(left, mid);
      mergeSortRecursive(mid + 1, right);
      merge(left, mid, right);
    }

    function merge(left: number, mid: number, right: number): void {
      const leftArr = array.slice(left, mid + 1);
      const rightArr = array.slice(mid + 1, right + 1);
      let i = 0;
      let j = 0;
      let k = left;

      while (i < leftArr.length && j < rightArr.length) {
        // Compare step
        steps.push({
          type: 'compare',
          array: [...array],
          highlightedIndices: [left + i, mid + 1 + j],
          secondaryIndices: Array.from({ length: right - left + 1 }, (_, idx) => left + idx),
          sortedIndices: [...sortedIndices],
          pseudocodeLine: 7,
          description: `Comparing ${leftArr[i]} and ${rightArr[j]}`,
        });

        if (leftArr[i] <= rightArr[j]) {
          array[k] = leftArr[i];
          i++;
        } else {
          array[k] = rightArr[j];
          j++;
        }

        // Show merge placement
        steps.push({
          type: 'swap',
          array: [...array],
          highlightedIndices: [k],
          secondaryIndices: Array.from({ length: right - left + 1 }, (_, idx) => left + idx),
          sortedIndices: [...sortedIndices],
          pseudocodeLine: leftArr[i - 1] !== undefined && i > 0 && array[k] === leftArr[i - 1] ? 8 : 9,
          description: `Place ${array[k]} at position ${k}`,
        });

        k++;
      }

      while (i < leftArr.length) {
        array[k] = leftArr[i];
        steps.push({
          type: 'insert',
          array: [...array],
          highlightedIndices: [k],
          secondaryIndices: Array.from({ length: right - left + 1 }, (_, idx) => left + idx),
          sortedIndices: [...sortedIndices],
          pseudocodeLine: 10,
          description: `Copy remaining element ${leftArr[i]} to position ${k}`,
        });
        i++;
        k++;
      }

      while (j < rightArr.length) {
        array[k] = rightArr[j];
        steps.push({
          type: 'insert',
          array: [...array],
          highlightedIndices: [k],
          secondaryIndices: Array.from({ length: right - left + 1 }, (_, idx) => left + idx),
          sortedIndices: [...sortedIndices],
          pseudocodeLine: 10,
          description: `Copy remaining element ${rightArr[j]} to position ${k}`,
        });
        j++;
        k++;
      }

      // If this merge covers the full array, mark all as sorted
      if (left === 0 && right === array.length - 1) {
        for (let idx = left; idx <= right; idx++) {
          sortedIndices.push(idx);
        }
        steps.push({
          type: 'sorted',
          array: [...array],
          highlightedIndices: Array.from({ length: right - left + 1 }, (_, idx) => left + idx),
          secondaryIndices: [],
          sortedIndices: [...sortedIndices],
          pseudocodeLine: 5,
          description: `Merged subarray [${left}..${right}] is now sorted`,
        });
      } else {
        steps.push({
          type: 'pass-complete',
          array: [...array],
          highlightedIndices: Array.from({ length: right - left + 1 }, (_, idx) => left + idx),
          secondaryIndices: [],
          sortedIndices: [...sortedIndices],
          pseudocodeLine: 5,
          description: `Merged subarray [${left}..${right}]`,
        });
      }
    }

    mergeSortRecursive(0, array.length - 1);

    return steps;
  },
};
